docs/Configuration/Dialplan/Pattern-Matching.md

```diff
@@ -3,331 +3,249 @@ title: Pattern Matching
 
-X
--
+Character Set
+-------------
 
-The letter **X** or **x** represents a single digit from 0 to 9.
+Place the characters you want to match between square brackets. If you want to match an `a` or an `e`, use `[ae]`, like in `gr[ae]y`, which matches either `gray` or `grey`.
 
-Z
--
+**Examples**:
+* `_gr[ae]y`: Matches `gray`, `grey`
+* `_[1-468]`: Matches `1`, `2`, `3`, `4`, `6`, `8`. Note that it does **NOT** match numbers from one to four hundred sixty-eight!
+* `_a[c-e]`: Matches `ac`, `ad`, `ae`
+* `_[127-9ac-e]`: Matches `1`, `2`, `7`, `8`, `9`, `a`, `c`, `d`, `e`
 
-The letter **Z** or **z** represents any digit from 1 to 9.
 
-N
--
+Special Symbols
+---------------
 
-The letter **N** or **n** matches any digit from 2-9.
+Some Character Sets can be written in a short form. Here are the most common ones:
 
+* `_[0-9]` = `_X` (or `x`): Represents a single digit from 0 to 9.
+* `_[1-9]` = `_Z` (or `z`): Represents any digit from 1 to 9.
+* `_[2-9]` = `_N` (or `n`): Matches any digit from 2 to 9.
 
+**Examples**:
+* `_6[34]XX`: Short form of `_64XX` and `_63XX`, matches `6400`, `6300`, `6401`, `6301`, `6420`, `6320`, `6433`, `6333`, ...
+* `_6[34X]`: Matches `63`, `64`, `6X`. The `X`, `N`, and `Z` convenience notations have no special meaning within a set, they represent the literal character.
 
-Ranges used in Pattern Matching
-===============================
-
-A numeric range can be used to match against a dialed number. This is also called a Character Set
-
+!!! note
+    The only characters with special meaning within a set are the `-` character, to define a range between two characters, the  `\` character to escape a special character available within a set, and  
+[//]: # (end-note)
 
 
-[1237-9]
+Wilcards
 --------
 
-This pattern matches any digit or letter in the brackets. In this example, the pattern will match: 1,2,3,7,8,9
-
-
-
-Wilcards used in Pattern Matching
-=================================
-
-The following special characters are considered wildcards
+Within Asterisk patterns, we can also use other characters to represent a *family* of dial numbers. The following special characters are considered wildcards:
 
-* . The '.' character matches one or more characters
-* ! The '!' character matches zero or more characters immediately
-
-
-
-The exclamation mark wildcard (!), behaves specially and will be further explained below in 'Other Special Characters' below.
-
-Please make sure to read 'Be Careful With Wildcards in Pattern Matches' below.
+* `!`: Matches zero or more characters *immediately*. Equivalent to RegEx's `*` quantifier. This wildcard behaves specially. It is used in overlap dialing to dial through Asterisk. For example, `_9876!` would match any number that began with `9876` including `9876`, and would respond that the number was complete as soon as there was an unambiguous match.
+* `.`: Matches one or more characters. Equivalent to RegEx's `+` quantifier.
```

Comment:
1. Only `!` is used for overlap dialing or `.` can also be used?
2. I couldn't find nothing regarding overlap dialing on the Asterisk docs. Does this page exist? We could link it here. If it's missing, we can write something in that matter ([example](https://www.3cx.com/blog/voip-howto/overlap-dialing/))

Reply:
@InterLinked1 Can you give your thoughts on this since you probably have insight into overlap dialing?

Reply:
I don't think there's anything special dialplan wise for overlap dialing. It's not really an Asterisk concept, it's a standard thing in industry. This is something that exists for both ISDN and SIP, but in a SIP context, it allows a phone to not have a digit map configured locally corresponding to the dialplan in Asterisk, and instead it just sends an INVITE after each digit and Asterisk sends a 484 or 486 (I think) if the destination doesn't match an extension. In the case of simple switch, it's actually very direct, sig_analog just calls the PBX functions to see if an unambiguous match was been reached yet (or an ambiguous one, with a timeout).

Basically, it makes it work similarly to native dialing in Asterisk, like with chan_dahdi's simple switch, or the DISA application, etc. It will start executing an extension if that is the only unambiguous match that remains.

If you have a pattern like `_X!` for example, that would match almost anything, so that would *break* overlap dialing really, so typically you don't want to use broad patterns like that in the incoming context for an endpoint. However, I do use them all the time in other contexts that are "in the middle", like for routing purposes, where you just want to match anything that goes through a context.

So to answer your question, you typically don't use wildcard pattern matches for the incoming context since that makes overlap dialing impossible, since anything would immediately match rather than the specific things that should be valid. These are useful, but not for overlap dialing per se, more just for pattern matching in general, where you might want to watch things starting with something, regardless of what comes after. So they really can be though of just in terms of their regex equivalents.

For example, `_X!` will match any numeric extension, although `_X.` would not match single-digit extensions.

Neither of these would match the extension, say, `#`, but something like `_[sA-D0-9*#]!` would (this is my default pattern match I use to catch *everything*).

Sorry for rambling, hope that answered the question but if not... I can clarify.